### ISSUE TITLE
Bug in ElasticSearch and Spark SQL: Using SQL to query out data from JSON documents is totally wrong!

### ISSUE THREAD

**Issue by arisvlasakakis-ck** [elastic/elasticsearch-hadoop#377]
I'm using ElasticSearch with elasticsearch-spark-BUILD-SNAPSHOT and Spark/SparkSQL 1.2.0, from Costin Leau's advice.

I want to query ElasticSearch for a bunch of JSON documents from within SparkSQL, and then use a SQL query to simply query for a column, which is actually a JSON key -- normal things that SparkSQL does using the SQLContext.jsonFile(filePath) facility. The difference I am using the ElasticSearch container.

The big problem: when I do something like 

SELECT jsonKeyA from tempTable;

I actually get the WRONG KEY out of the JSON documents! I discovered that if I have JSON keys physically in the order D, C, B, A in the json documents, the elastic search connector discovers those keys BUT then sorts them alphabetically as A,B,C,D - so when I SELECT A from tempTable, I actually get column D (because the physical JSONs had key D in the first position). This only happens when reading from elasticsearch and SparkSQL.

It gets much worse: When a key is missing from one of the documents and that key should be NULL, the whole application actually crashes and gives me a java.lang.IndexOutOfBoundsException -- the schema that is inferred is totally screwed up. 

In the above example with physical JSONs containing keys in the order D,C,B,A, if one of the JSON documents is missing the key/column I am querying for, I get that java.lang.IndexOutOfBoundsException exception.

I am using the BUILD-SNAPSHOT because otherwise I couldn't build the elasticsearch-spark project, Costin said so.

Any clues here? Any fixes?

**Comment by costin** [comment]
@arisvlasakakis-ck 
I've tried reproducing the issues and couldn't.  Here's a code snippet of the test:
```scala
      val target = "spark-test/scala-sql-varcols"
      
      val trip1 = Map("reason" -> "business", "airport" -> "SFO")
      val trip2 = Map("participants" -> 5, "airport" -> "OTP")

      sc.makeRDD(Seq(trip1, trip2)).saveToEs(target)
      
      val schemaRDD = sqc.sql("CREATE TEMPORARY TABLE sqlvarcol " + 
              "USING org.elasticsearch.spark.sql " +
              "OPTIONS (resource '" + target + "')");
      
      val allResults = sqc.sql("SELECT * FROM sqlvarcol")
      assertEquals(2, allResults.count())
      println(allResults.schemaString)

      val filter = sqc.sql("SELECT * FROM sqlvarcol WHERE airport = 'OTP'")
      assertEquals(1, filter.count())
      
      val nullColumns = sqc.sql("SELECT reason, airport FROM sqlvarcol ORDER BY airport")
      val rows = nullColumns.take(2)
      assertEquals("[null,OTP]", rows(0).toString())
      assertEquals("[business,SFO]", rows(0).toString())
```

Basically I'm adding two documents, with one field in common and one different. First, I check the schema which is returned properly:
``
root
 |-- airport: string (nullable = true)
 |-- participants: long (nullable = true)
 |-- reason: string (nullable = true)
``
Then the test selects all the columns and checks the matches - again everything is in order. Last but not least, two columns (one in both docs, one only in one) are selected and ordered - as you can see, both checks are fine and the missing column is returned as null. Notice how the ordering of the columns within the documents is respected.

I suggest creating a small example or sharing your test code through a [gist](gist.github.com) and pasting the link back.

P.S. > I am using the BUILD-SNAPSHOT because otherwise I couldn't build the elasticsearch-spark project, Costin said so.

Actually what I said what that Spark 1.2 broke backwards compatibility with 1.2 and thus es-hadoop 2.1.0.Beta3, which was released before, will only work with Spark 1.1.
If you have issues building the project do raise a separate issue _after_ reading the readme. The project is built daily on two CI servers (and on my machine) so likely there's a misconfiguration on your part.

**Comment by costin** [comment]
Without any more update, I'm closing the issue. If the problem reappears, make sure to get the latest Beta and please open up a new issue.
